pkg/piecestore/rpc/client/readerwriter.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package client
+
+import (
+	"fmt"
+	"log"
+
+	pb "storj.io/storj/protos/piecestore"
+)
+
+// StreamWriter -- Struct for writing piece to server upload stream
+type StreamWriter struct {
+	stream pb.PieceStoreRoutes_StoreClient
+}
+
+// Write -- Write method for piece upload to stream
+func (s *StreamWriter) Write(b []byte) (int, error) {
+	if err := s.stream.Send(&pb.PieceStore{Content: b}); err != nil {
+		return 0, fmt.Errorf("%v.Send() = %v", s.stream, err)
+	}
+
+	return len(b), nil
+}
+
+// Close -- Close Write Stream
+func (s *StreamWriter) Close() error {
+	reply, err := s.stream.CloseAndRecv()
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Route summary: %v", reply)
+
+	return nil
+}
+
+// StreamReader -- Struct for reading piece download stream from server
+type StreamReader struct {
+	stream       pb.PieceStoreRoutes_RetrieveClient
+	overflowData []byte
+}
+
+// Read -- Read method for piece download stream
+func (s *StreamReader) Read(b []byte) (int, error) {
+	if len(s.overflowData) > 0 {
+		n := copy(b, s.overflowData)
+		s.overflowData = s.overflowData[:len(s.overflowData)-n]
```

Comment:
oh, sorry, shoot, i missed this last time. this does hold up the merge, and i suspect this implies this isn't covered by unit tests (but getting this covered by unit tests shouldn't hold up the merge)

here's how to add one line to fix your change:

```golang
if len(s.overflowData) > 0 {
   n := copy(b, s.overflowData)
   copy(s.overflowData, s.overflowData[n:]) // important! slide the remaining data to the beginning of the buffer
   s.overflowData = s.overflowData[:len(s.overflowData)-n]
   return n, nil
}
```

or instead, below makes more sense in this case (though not every case! sometimes this second strategy dramatically increases garbage collection pressure, but not this time, since the s.overflowData buffer isn't reused anyway):

```golang
if len(s.overflowData) > 0 {
   n := copy(b, s.overflowData)
   s.overflowData = s.overflowData[n:]
   return n, nil
}
```

Reply:
I meant to do what you put ` s.overflowData = s.overflowData[n:]` but I think I mixed self up and did it backwards. whoops!